Simpler and more precise border production

raggi opened the pull request:
This includes three changes that could be included independently though each has some side effect that is adjusted by the latter to better align with the overall goal.

I set out to remove the inconsistent gap and miscoloration around windows. This was originating in the interactions with DWM decorations.

First the API that we're using to read window size is shifted to one that returns Window dimensions without including the DWM/theme decorations.

Next in order to fix a number of new regressions vs. the application database, we needed to remove the use of the invisible borders - these were a side effect of the DWM decorations. They're no longer needed, as application Window dimensions are now correct, with one notable interaction with theming (below).

Finally we move away from allowing/using DWM decorations on the border window. While this was giving us a rounded corner, it was causing offset bugs due to the fact that the theme is not symmetrical. This was making it impossible to draw very thin borders correctly, for example few pixel borders.

There is one final caveat noted here, the problem exists already in a different form, but for the same as yet unresolved reasons. Single pixel borders work correctly with this setup but there remain overlapping concerns with the system them. The windows DWM theme adds an active window border as well, and how opaque this border is depends upon the theme (this is present even if the drop shadow is off). That border often produces what appears to be a single semi-transparent pixel border around the window that looks like a "gap" in many scenarios. Most users will want to set border-offset to -1 and then increase the border with by 1 to get a solid active window border color underneath the whole border edge, without bleeding through to the background or similar, that otherwise looks like a gap. This leaves what has the look and feel of a single pixel border in many scenarios except as follows.

As an example the Epic launcher uses a different DWM window style, that adjusts the border behavior. The border there is completely opaque and spills over a single pixel border window, effectively making the active border window invisible. For users trying to do that single pixel border style, it will effectively disappear under those windows. The correct fix for this is to interrogate the DWM border style of the windows we are decorating and adjust for this. As it stands right now I doubt any users are affected by this, as if they're arranging for single pixel borders today, they will be missing borders on many applications that have incorrect sizing and offsetting.

One final thing that is also worth considering for a near future, aside from detecting DWM styles for border edges, is to also use that to apply a small fixup on the window corner radius for those applications. Relatedly, we could also offer a general border radius option, and a really polished fix would work on addressing DWM properties that observe the default Windows 11 corner styles, where the bottom right "resizable border" has a non-cicular inset on the corner curve - we could fill that in, or conform to it (the latter requires doing some more fancy drawing).

LGUG2Z commented:
I haven't looked through the code yet, but some initial thoughts from trying out the branch:

The defaults set for border width and offset with this branch result in `yuge` borders; we should probably set these defaults in the code to `10` for width and `-3` for offset get a "look" similar to the defaults we have in 0.1.21.

![image](https://github.com/LGUG2Z/komorebi/assets/13164844/e4da85b2-79b5-4a13-b7ff-65eb22603d85)


```json
{
  "default_workspace_padding": 10,
  "default_container_padding": 20
}
```

With these settings on v0.1.21 the gaps in in between my containers are smaller than they are on this branch

v0.1.21 here
![Screenshot 2024-02-20 181131](https://github.com/LGUG2Z/komorebi/assets/13164844/6a0558b9-c06a-40e3-915c-74d5a29ea950)

raggi/borders here
![Screenshot 2024-02-20 180958](https://github.com/LGUG2Z/komorebi/assets/13164844/5aa11460-56e8-4a73-a738-30ffbc3194f5)

LGUG2Z commented:
I haven't looked through the code yet, but some initial thoughts from trying out the branch:

The defaults set for border width and offset with this branch result in `yuge` borders; we should probably set these defaults in the code to `10` for width and `-3` for offset get a "look" similar to the defaults we have in 0.1.21.

![image](https://github.com/LGUG2Z/komorebi/assets/13164844/e4da85b2-79b5-4a13-b7ff-65eb22603d85)


```json
{
  "default_workspace_padding": 10,
  "default_container_padding": 20
}
```

With these settings on v0.1.21 the gaps in in between my containers are smaller than they are on this branch

v0.1.21 here
![Screenshot 2024-02-20 181131](https://github.com/LGUG2Z/komorebi/assets/13164844/6a0558b9-c06a-40e3-915c-74d5a29ea950)

raggi/borders here
![Screenshot 2024-02-20 180958](https://github.com/LGUG2Z/komorebi/assets/13164844/5aa11460-56e8-4a73-a738-30ffbc3194f5)

LGUG2Z commented:
Another photo with all padding set to `0` on this branch which shows extra padding around the windows that isn't present in v0.1.21 (compare with the gaps when setting everything to `0` on this video: https://www.youtube.com/watch?v=6QYLao953XE)

![image](https://github.com/LGUG2Z/komorebi/assets/13164844/e2d6cc25-0bc4-496b-8653-89b23bf67603)

raggi commented:
> Another photo with all padding set to 0 on this branch which shows extra padding around the windows that isn't present in v0.1.21 (compare with the gaps when setting everything to 0 on this video: https://www.youtube.com/watch?v=6QYLao953XE)

Interesting, this looks like we maybe take border size into account when calculating padding?

raggi commented:
@LGUG2Z please take a look. I've added two more changes here:

- Window decorations are now correctly accounted for in window reposition, so that the target size is what the window content size becomes.
- Window sizes are now adjusted to account for border offset and size, so the border is always visible.
- The border window is always made topmost, so the complete border is visible and non-shaded such as when overlapping with a default DWM alpha blend 1px border, and also when there is a non-DWM custom border.

With this setup you can set offset=-1 width=1 with 0 padding to get a perfect gapless grid.

Please let me know what you want the default border width to be. I would recommend a default border offset of -1, so that the in-window-bounds translucent DWM border always has a solid color underneath. The visible border width will now be exactly the number of pixels specified in `width`, so we just need a pixel sized expectation for this size.

LGUG2Z commented:
Hmm possibly, I wonder if this is a result of removing the "invisible" border modifications in `window.rs`? 🤔

raggi commented:
I sent another commit that adjust the default to -1 offset, 8px width, I think this feels close to the original intent.

LGUG2Z commented:
Awesome! Can't wait to try this when I'm back home 🤞
